### PR TITLE
OSD: add get_mapped_pools asok command

### PR DESCRIPTION
**Pull request by xiaoxichen**
get_mapped_pools will return all pool_id that
mapped to the paticular OSD.

It is super useful for monitoring system that want to
**group** OSDs based on application. PoolID is a nature
grouping tag to be attached to OSDs.

Signed-off-by: Xiaoxi Chen <xiaoxchen@ebay.com>

### CODE REVIEW

File: src/osd/OSD.h [hunk 1]
```diff
@@ -1944,6 +1944,14 @@ class OSD : public Dispatcher,
     return pg_map.size();
   }
 
+  std::set<int> get_mapped_pools() {
```
Comment (tchaikov) [inline src/osd/OSD.h]: could you define this method in `OSD.cc`?

File: src/osd/OSD.h [hunk 2]
```diff
@@ -1944,6 +1944,14 @@ class OSD : public Dispatcher,
...
+    std::set<int> pools;
```
Comment (tchaikov) [inline src/osd/OSD.h]: `std::vector` would suffice.

File: src/osd/OSD.cc [hunk 3]
```diff
@@ -2292,6 +2292,19 @@ will start to track new ops received afterwards.";
     f->open_object_section("compact_result");
     f->dump_float("elapsed_time", duration);
     f->close_section();
+  } else if (admin_command == "get_mapped_pools") {
+    f->open_object_section("mapped_pools");
+    set<int> poollist = get_mapped_pools();
+
+    std::stringstream buffer;
+    for (set<int>::iterator it = poollist.begin();
```
Comment (tchaikov) [inline src/osd/OSD.cc]: please see `types.h`, we have helpers for this purpose in our arsenal.

File: src/osd/OSD.cc [hunk 4]
```diff
@@ -2822,6 +2835,10 @@ void OSD::final_init()
                                      " WARNING: Compaction probably slows your requests");
   assert(r == 0);
 
+  r = admin_socket->register_command("get_mapped_pools", "get_mapped_pools",
+                                     asok_hook,
+                                     "dump pools mapped(at least one active PG on) to this OSD.");
```
Comment (tchaikov) [inline src/osd/OSD.cc]: add a space before "(". moreover, could you rephrase this help message? i think

> at least one active PG on

in the parentheses is confusing and hard to read. what is "active" PG in this context? probably we can change it to something like

> dump pools whose PG(s) are mapped to this OSD

?

File: src/osd/OSD.cc (hunk 3 above)
Comment (xiaoxichen) [reply src/osd/OSD.cc]: addressed, thanks

File: src/osd/OSD.h (hunk 1 above)
Comment (xiaoxichen) [reply src/osd/OSD.h]: done

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2017-11-28_10:23:40-rados-wip-kefu-testing-2017-11-28-1542-distro-basic-mira/

failures tracked/fixed by

- http://tracker.ceph.com/issues/22093
- http://tracker.ceph.com/issues/22043
- #19225

**Comment by xiaoxichen** [comment]
@tchaikov can I backport it to luminous?
